### PR TITLE
22 permanent errors

### LINKED ISSUE #22: Abort processes for permanent errors
- couchbackup/couchrestore version(s) that are affected by this issue.
>1.4.0
- Node.js version.
>4.8.2
- A sample of the code or command line statement that demonstrates the issue.

Following on from #17 there are other cases where the tool will persist in streaming a large file with no hope of success. For example the case of attempting a restore without `_writer` permission for the database.

There are also cases where this applies to the backup process, for example, having `_reader` permissions revoked (if `_reader` was not there in the first place it would never spool the `_changes` and terminate before the batch get stage where the process currently tries forever to complete.

### PR DESCRIPTION
**Pull request by ricellis**
Thanks for your hard work, please ensure all items are complete before opening a PR.

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](https://github.com/cloudant/couchbackup/blob/master/DCO1.1.txt)
- [x] You have [added tests](https://github.com/cloudant/couchbackup/blob/master/CONTRIBUTING.md#testing) for any code changes
- [x] You have updated the [CHANGES.md](https://github.com/cloudant/couchbackup/blob/master/CHANGES.md)
- [x] You have completed the PR template below:

## What

Fixed issues with backup and restore not terminating or issue correct exit codes for various permanent errors.

## How

* Consolidated the HTTP response error handling.
* Added `HTTPError` and `HTTPFatalError` classes of `BackupError` and added `isFatal` flag to more easily identify fatal errors.
* Made a default of 4XX except 429 fatal and 429 & >=500 transient.
* Special cased auth errors with `err.name` to get unique exit codes.
* Updated readme for auth errors to be for both backup and restore.
* Terminated async queues on fatal error conditions.
* Removed event listeners when calling `callback(error)` (i.e. ending the process on an error) to avoid the chance of multiple calls to the `callback` from parallel fatal errors.
* Ensured that after doing a `callback(error)` code blocks used `return` and didn't fall.
* Added callbacks for `writer` queue tasks to propagate errors.
* Called `done(err)` for `writer._transform` on queue task errors.
* Added debug statements to writer.

## Testing

Added `fatalerrors.js` to test termination and exit codes for API and CLI cases.
* Backup
    * should terminate on BulkGetError
    * should terminate on IncompleteChangesInLogFile
    * should terminate on LogDoesNotExist
    * should terminate on NoLogFileName
    * should terminate on SpoolChangesError
    * should terminate on Unauthorized _bulk_get check
    * should terminate on _bulk_get HTTPFatalError
    * should terminate on _changes HTTPFatalError
    * terminate on Forbidden no _reader
* Restore
    * should terminate on Forbidden no _writer
    * should terminate on RestoreDatabaseNotFound
    * should terminate on Unauthorized db existence check
    * should terminate on _bulk_docs HTTPFatalError

Updated `test/request.js` to use `404` for fatal error, added additional tests for non-fatal and fatal callbacks.
Fixed `test/app.js` to test a boolean type for `resume` without needing an extra `log` parameter to be specified.

## Issues

Fixes #22

### CODE REVIEW

File: test/fatalerrors.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    before('Set process data for test', function() {
+      // Copy env and argv so we can reset them after the tests
+      processEnvCopy = JSON.parse(JSON.stringify(process.env));
+
+      // Set up a proxy to point to our nock server because the nock override
+      // isn't visible to the spawned CLI process
+      if (!params.useApi) {
+        proxy = httpProxy.createProxyServer({target: url}).listen(8888, 'localhost');
+      //   proxy.on('open', done);
```
Comment (tomblench) [inline test/fatalerrors.js]: commented code - should it be removed?

File: test/fatalerrors.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+      }
+
+      // setup environment variables
+      process.env.COUCH_URL = (params.useApi) ? url : 'http://localhost:8888';
+    });
+
+    after('Reset process data', function() {
+      process.env = processEnvCopy;
+      if (!params.useApi) {
+        // proxy.on('close', done);
```
Comment (tomblench) [inline test/fatalerrors.js]: commented code - should it be removed?

File: test/request.js [hunk 3]
```diff
@@ -19,10 +19,10 @@ var assert = require('assert');
 var nock = require('nock');
 var request = require('../includes/request.js');
 
-describe('#unit Check request response', function() {
-  const url = 'http://localhost:5984';
-  const client = request.client(url, 1);
+const url = 'http://localhost:5984';
```
Comment (tomblench) [inline test/request.js]: Will running nock on port 5984 be a problem if we ever want to run in an environment with a local couch running?

File: app.js [hunk 4]
```diff
@@ -122,7 +145,7 @@ module.exports = {
       targetStream.write('\n');
     }
 
-    backup(srcUrl, opts.bufferSize, opts.parallelism, opts.log, opts.resume)
+    const internalEE = backup(srcUrl, opts.bufferSize, opts.parallelism, opts.log, opts.resume)
```
Comment (ricellis) [inline app.js]: Fixed in 59ab59e90ff7bec4d06c241907001cf821f306cb

File: includes/backup.js [hunk 5]
```diff
@@ -83,18 +83,16 @@ module.exports = function(dbUrl, blocksize, parallelism, log, resume) {
...
-      default:
-        request.checkResponseAndCallbackFatalError(res, function(err) {
-          callback(err);
-        });
-    }
+    request.checkResponseAndCallbackError(res, callback, function(res) {
+      switch (res.statusCode) {
+        case 404:
+          return new error.BackupError('BulkGetError', 'ERROR: Database does not support /_bulk_get endpoint');
+        case 405:
```
Comment (ricellis) [inline includes/backup.js]: Fixed in 6df5dd0b6abad9a16df57c9eea909ae22fedd0b7

File: includes/writer.js [hunk 6]
```diff
@@ -45,31 +42,43 @@ module.exports = function(couchDbUrl, bufferSize, parallelism, ee) {
...
+      }
       if (err) {
-        writer.emit('error', err);
+        debug(`Error writing docs ${err.name} ${err.message}`);
+        if (err.isFatal) {
+          debug(`Fatal error: ${err.name}`);
+          response.abort();
+          q.kill();
+          cb(err);
+          return;
```
Comment (ricellis) [inline includes/writer.js]: Fixed in 3716068e5692578c0a904c403508a5a567ec99cd

File: test/fatalerrors.js (hunk 1 above)
Comment (ricellis) [inline test/fatalerrors.js]: Fixed in 47d149268227e40dced7b8d16483e9a84b24f37e

File: test/fatalerrors.js (hunk 2 above)
Comment (ricellis) [inline test/fatalerrors.js]: Fixed in 47d149268227e40dced7b8d16483e9a84b24f37e

File: test/request.js (hunk 3 above)
Comment (ricellis) [inline test/request.js]: Could be, changed in 35a873b3ea83c20731f1ffee0891391fa74c7e97
Also added a `beforeEach` to clean the mocks which I think will help for failure cases as I did in the fatalerrors test.

**Pull request merged**